src/core/comm/coll.cc

```diff
@@ -102,24 +102,24 @@ int collCommCreate(CollComm global_comm,
   global_comm->mpi_comm_size_actual = 1;
   global_comm->mpi_rank             = 0;
   if (global_comm->global_rank == 0) {
-    pthread_barrier_init((pthread_barrier_t*)&(thread_comms[global_comm->unique_id].barrier),
+    pthread_barrier_init((pthread_barrier_t*)&(thread_comms[global_comm->unique_id]->barrier),
                          nullptr,
                          global_comm->global_comm_size);
-    thread_comms[global_comm->unique_id].buffers =
+    thread_comms[global_comm->unique_id]->buffers =
       (const void**)malloc(sizeof(void*) * global_comm_size);
-    thread_comms[global_comm->unique_id].displs =
+    thread_comms[global_comm->unique_id]->displs =
       (const int**)malloc(sizeof(int*) * global_comm_size);
     for (int i = 0; i < global_comm_size; i++) {
-      thread_comms[global_comm->unique_id].buffers[i] = nullptr;
-      thread_comms[global_comm->unique_id].displs[i]  = nullptr;
+      thread_comms[global_comm->unique_id]->buffers[i] = nullptr;
+      thread_comms[global_comm->unique_id]->displs[i]  = nullptr;
     }
     __sync_synchronize();
-    thread_comms[global_comm->unique_id].ready_flag = true;
+    thread_comms[global_comm->unique_id]->ready_flag = true;
   }
   __sync_synchronize();
-  volatile ThreadComm* data = &(thread_comms[global_comm->unique_id]);
-  while (data->ready_flag != true) { data = &(thread_comms[global_comm->unique_id]); }
-  global_comm->comm = &(thread_comms[global_comm->unique_id]);
```

Comment:
The reason of this PR is because of this line. When the thread_comms vector get resized, the address is changed.

Reply:
yeah and even if the address was stable, it'd be bad for performance to copy the whole structs.